File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/ValidationRunData.kt
package net.nemerosa.ontrack.model.structure

/**
 * Data associated with a [ValidationRun] and
 * a [ValidationDataType] by ID.
 *
 * @param T Type of data
 * @param descriptor Type descriptor
 * @param data Data
 */
class ValidationRunData<out T>(
        val descriptor: ValidationDataTypeDescriptor,
        val data: T
)

fun <C, T> ValidationDataType<C, T>.data(data: T) =
        ValidationRunData(
                descriptor,
                data
        )

File: ontrack-web-core/components/framework/validation-run-data/ValidationRunData.js
import {Dynamic} from "@components/common/Dynamic";

export default function ValidationRunData({data}) {
    const shortTypeName = data ? data.descriptor.id.slice("net.nemerosa.ontrack.extension.".length) : ''

    if (shortTypeName) {
        const actualData = data.data
        let props = {}
        if (typeof actualData === 'object') {
            props = {...actualData}
        } else {
            props = {value: actualData}
        }

        return <Dynamic
            path={`framework/validation-run-data/${shortTypeName}`}
            props={props}
        />
    } else {
        return ''
    }
}